release/alpa_tests/train_opt_2_7b_minimum.py

```diff
@@ -0,0 +1,477 @@
+        data_parallel=-1,
+        operator_parallel=training_args.operator_parallel,
+        pipeline_parallel=training_args.pipeline_parallel,
+    )
+
+    p_train_step = alpa.parallelize(train_step, method=method, donate_argnums=(0,))
+
+    logger.info("***** Training *****")
+    logger.info(f"  Num examples = {len(train_dataset)}")
+    logger.info(f"  Num Epochs = {num_epochs}")
+    logger.info(
+        "  Batch size per device (w. accumulation) = "
+        f"{training_args.per_device_train_batch_size}"
+    )
+    logger.info(
+        f"  Global train batch size (w. parallel & distributed) = {train_batch_size}"
+    )
+    logger.info(f"  Total optimization steps = {total_train_steps}")
+    logger.info(f"  NCCL mode = {global_config.nccl_mode}")
+
+    train_time = 0
+    train_metrics = []
+    epochs = tqdm(range(num_epochs), desc="Epoch ... ", position=0)
+
+    step_ct = 0
+    last_time = time.time()
+
+    epochs.write("Initial compilation. This might take some minutes...")
+
+    for epoch in epochs:
+        train_start = time.time()
+
+        # Generate an epoch by shuffling sampling indices from the train dataset
+        train_loader = data_loader(train_dataset, train_batch_size, shuffle=True)
+        steps_per_epoch = len(train_dataset) // train_batch_size
+
+        for step in tqdm(
+            range(steps_per_epoch), desc="Training...", position=1, leave=False
+        ):
+            batch = next(train_loader)
+            batch["position_ids"] = (
+                batch["attention_mask"].cumsum(axis=1) * batch["attention_mask"]
+            ) - 1
+            state, train_metric = p_train_step(state, batch)
+            train_metrics.append(train_metric)
+
+            cur_step = epoch * (len(train_dataset) // train_batch_size) + step
+
+            step_ct += 1
+            if cur_step % training_args.logging_steps == 0 and cur_step > 0:
+                latency = (time.time() - last_time) / step_ct
+                throughput_tokens = np.prod(batch["input_ids"].shape) / latency
+                throughput_tflops = alpa.util.compute_gpt_tflops(
+                    batch_size=batch["input_ids"].shape[0],
+                    seq_len=batch["input_ids"].shape[1],
+                    num_layers=config.num_hidden_layers,
+                    hidden_size=config.hidden_size,
+                    vocab_size=config.vocab_size,
+                    num_gpus=alpa.get_global_num_devices(),
+                    latency=latency,
+                )
+                step_ct = 0
+
+                # Save metrics
+                train_time += time.time() - train_start
+
+                train_metric = jax.tree_map(np.mean, train_metric)
+
+                epochs.write(
+                    f"Step... {cur_step} | "
+                    f"Loss: {train_metric['loss'].mean():.4f}, "
+                    f"Throughput: {throughput_tokens:.2f} token/s, "
+                    f"{throughput_tflops:.2f} TFLOP/s"
```

Comment:
do we continuously track this metrics in dashboard ? And generally speaking we should throw away first 1~2 steps and average the subsequent ones for perf tracking 

not blocking and can be next PR